_posts/2020-11-12-symbolic-execution-with-ds-test.md

```diff
@@ -0,0 +1,429 @@
+- All storage slots are set to zero at the beginning of the tests (meanings tests are effectively run against an empty blockchain)
+
+In fact, the only symbolic variables introduced into the test environment are those that are
+specified in the signature of the test method. This means that the proofs are exhaustive *only over
+the input variables*. As an example, consider the `prove_transfer` test from the example above. The
+`totalSupply` is always `uint(-1)`, and the test would not catch obviously faulty implementations of
+`transfer` like the one below:
+
+```solidity
+function transfer(address dst, uint amt) public {
+    require(totalSupply > uint56(-1), "whoops");
+    balanceOf[msg.sender] = sub(balanceOf[msg.sender], amt);
+    balanceOf[dst]        = add(balanceOf[dst], amt);
+}
+```
+
+## Execution Against Mainnet State
+
+`hevm` allows us to fetch state from an rpc node, and we can also use this to write symbolic tests
+against mainnet state. As an example, lets run `prove_transfer` against the [balancer
+token](https://etherscan.io/address/0xba100000625a3754423978a60c9317c58a424e3D#code):
+
+```solidity
+interface ERC20 {
+    function balanceOf(address account) external view returns (uint256);
+    function transfer(address recipient, uint256 amount) external returns (bool);
+}
+
+contract TestBal is DSTest, SafeMath {
+    function prove_transfer(address dst, uint amt) public {
+        // BAL: https://etherscan.io/address/0xba100000625a3754423978a60c9317c58a424e3D#code
+        ERC20 token = ERC20(0xba100000625a3754423978a60c9317c58a424e3D);
+
+        // ignore cases where we don't have engough tokens
+        if (amt > bal.balanceOf(address(this))) return;
+
+        uint preBalThis = token.balanceOf(address(this));
+        uint preBalDst  = token.balanceOf(dst);
+
+        token.transfer(dst, amt);
+
+        // no change for self-transfer
+        uint delta = dst == address(this) ? 0 : amt;
+
+        // balance of `this` has been reduced by `delta`
+        assertEq(sub(preBalThis, delta), bal.balanceOf(address(this)));
+
+        // balance of `dst` has been increased by `delta`
+        assertEq(add(preBalDst, delta), bal.balanceOf(dst);
+    }
+}
+```
+
+Lets run this test as the [balancer DAO](https://etherscan.io/token/0xba100000625a3754423978a60c9317c58a424e3d?a=0xb618f903ad1d00d6f7b92f5b0954dcdc056fc533) to make sure that we have plenty of `BAL`:
+
+```
+$ DAPP_TEST_ADDRESS=0xb618f903ad1d00d6f7b92f5b0954dcdc056fc533 dapp test --rpc-url <URL>
+[FAIL] prove_transfer(address,uint256)
+
+Failure: prove_transfer(address,uint256)
+
+  Counter Example:
+
+    result:   Revert("ERC20: transfer to the zero address")
+    calldata: prove_transfer(0x0000000000000000000000000000000000000000, 0)
+```
+
+We have uncovered another edge case! The balancer token disallows transfers to the zero address.
+
+It is also worth noting that symbolic execution against rpc state can be significantly more
+performant than fuzzing via rpc.
```

Comment:
Why is that?

Reply:
because the fuzz tests fetch the state for each run. Although now I think about it again, there are options that allow this state to be cached locally. I'm just going to remove this line.

Reply:
Ah ok